### PR TITLE
fix(core,topology): Replace feed with send to fix flushing behaviour on disk buffers

### LINKED ISSUE #10806: Disk buffering doesn't send events until more come in
When using disk buffering, the last N log events are not delivered via the Loki sink until more events come in.

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Vector Version

```
vector 0.19.0 (x86_64-unknown-linux-gnu da60b55 2021-12-28)
```

#### sources/journal.toml

```toml
type = 'journald'
acknowledgements.enabled = false
batch_size = 1
current_boot_only = false
```

#### sinks/loki.toml

```toml
type = 'loki'
inputs = ['journal']
endpoint = 'https://loki.internal.home.arpa/'
labels.app = '{{src.app}}'
labels.file = '{{src.file}}'
labels.host = '{{src.host}}'
labels.src = '{{src.type}}'
encoding.codec = 'json'
auth.strategy = 'basic'
auth.user = '$LOKI_USER'
auth.password = '$LOKI_PASS'
batch.timeout_secs = 1
out_of_order_action = 'rewrite_timestamp'
buffer.type = 'disk'
buffer.max_size = 104900000
```

### Expected Behavior

Log events from the systemd journal are delivered to my Loki instance at `https://loki.internal.home.arpa/` within a few seconds of Vector reading them from the journal.

### Actual Behavior

With disk buffering, the last N events aren't delivered (seemingly at all) until more events are logged to the journal. For example, if I restart Vector (which itself is logging to the journal), I won't see any journal messages about the restart appear in Loki until the _second_ time I restart Vector... and then I will only see the first restart's events, not the second.

If I change the `buffer.*` options to those shown below, events are delivered more-or-less immediately.

```toml
buffer.type = 'memory'
buffer.max_events = 500
```

### Additional Context

Vector is running in Arch Linux in VMware Workstation. However I have seen the same thing happening on other hardware/OS combinations.

Loki is version 2.4.1 and running on Windows 10 Pro.

### PR DESCRIPTION
**Pull request by spencergilbert**
Signed-off-by: Spencer Gilbert <spencer.gilbert@datadoghq.com>

Closes #10806

Our transform concurrency work replaced a `SinkExt::send_all` call -- which will generate a `poll_flush call` -- with a `SinkExt::feed` call, which explicitly does not flush.

### CODE REVIEW

**Comment by spencergilbert** [comment]
@jszwedko likely want to pick this into our upcoming point release

File: lib/vector-core/src/transform/mod.rs [hunk 1]
```diff
@@ -239,7 +239,7 @@ impl TransformOutputs {
 
 async fn send_inner(buf: &mut Vec<Event>, output: &mut Fanout) {
     for event in buf.drain(..) {
-        output.feed(event).await.expect("unit error");
+        output.send(event).await.expect("unit error");
     }
```
Comment (lukesteensen) [inline lib/vector-core/src/transform/mod.rs:243]: Since this one is bounded, I think we'd be better off doing `feed` in the loop and then one `flush` at the end.

File: src/topology/builder.rs [hunk 2]
```diff
@@ -159,7 +159,7 @@ pub async fn build_pieces(
             let (mut fanout, control) = Fanout::new();
             let pump = async move {
                 while let Some(event) = rx.next().await {
-                    fanout.feed(event).await?;
+                    fanout.send(event).await?;
                 }
                 fanout.flush().await?;
```
Comment (lukesteensen) [inline src/topology/builder.rs:164]: Good catch on this one 😬

**Comment by jszwedko** [comment]
Nice find! Is there any way we could test this behavior from a higher layer?

**Comment by spencergilbert** [comment]
> Nice find! Is there any way we could test this behavior from a higher layer?

Credit to @tobz - my gut says probably not as it's a weird `Sink` footgun, but I'm not an expert here

**Comment by tobz** [comment]
Spencer's gut is right: it's tricky to test.  That said, we _could_ test this specific behavior, which is that if we wrap a debug-oriented `Sink` implementation, like literally just a mock that fulfills `Sink`, we could assert that it's actually calling `poll_flush` at the end.

You end up in the trap of figuring out whether or not you're actually using the logic, though.  Which is to say: we know that a certain piece of code is tested and functions correctly, but where do you test to assert that _that_ certain piece of code is what's actually being used?

Basically, we end up constructing something like a source/transform/disk buffered sink topology just to test it all out.  Is that the worst idea? Probably not, and it likely wouldn't even be our gnarly test in the repository... but it's just a bit clunky to build.

I'd be in support for such a test as a follow-on, since we know this fixes a regression, so long as all the existing unit/integration tests pass, etc.

File: lib/vector-core/src/transform/mod.rs (hunk 1 above)
Comment (tobz) [reply lib/vector-core/src/transform/mod.rs:243]: That's a good point. 👍🏻

Comment (spencergilbert) [reply lib/vector-core/src/transform/mod.rs]: Updated, better?

**Comment by lukesteensen** [comment]
I took at stab at writing a topology test for this, but it currently passes because it doesn't actually use disk buffers:

```diff
diff --git a/tests/topology.rs b/tests/topology.rs
index d8781971a..3ac99e93a 100644
--- a/tests/topology.rs
+++ b/tests/topology.rs
@@ -114,6 +114,36 @@ async fn topology_source_and_sink() {
     assert_eq!(vec![event], res);
 }

+#[tokio::test]
+async fn topology_flushes_on_idle() {
+    let event = Event::from("foo");
+
+    let (mut in1, source1) = source();
+    let transform1 = transform("bar", 0.0);
+    let (mut out1, sink1) = sink(10);
+
+    let mut config = Config::builder();
+    config.add_source("in1", source1);
+    config.add_transform("t1", &["in1"], transform1);
+    config.add_sink("out1", &["t1"], sink1);
+
+    let (topology, _crash) = start_topology(config.build().unwrap(), false).await;
+
+    in1.send(event).await.unwrap();
+
+    let res = tokio::time::timeout(Duration::from_secs(1), out1.next())
+        .await
+        .expect("timeout")
+        .map(into_message)
+        .expect("no output");
+    assert_eq!("foobar", res);
+
+    topology.stop().await;
+
+    let rest = out1.collect::<Vec<_>>().await;
+    assert_eq!(rest, vec![]);
+}
+
 #[tokio::test]
 async fn topology_multiple_sources() {
     let (mut in1, source1) = source();
```

I don't think it would be an enormous amount more work to have a test sink that includes a disk buffer, but agree it could be a followup.

**Review by tobz — APPROVED** [review]

**Comment by tobz** [comment]
Looking at the soak results, this feels somewhat expected to me given that there's now an extra operation happening after each resulting event from the transform is sent.

For transforms that are basically taking one event and spitting out a single event, now we've added a lot of overhead, relatively speaking.  Although, all things considered, the flush call should be a no-op with in-memory buffers, which is what these soak tests are using.

**Comment by spencergilbert** [comment]
I am saddened and confused :D

**Review by jszwedko — APPROVED** [review]

**Comment by jszwedko** [comment]
As mentioned offline the soak results are likely the artifact of this branch not being up-to-date with master.

**Review by lukesteensen — COMMENTED** [review]
Did we lose the change on the source side to `send` instead of `feed` in the pump task? I only see the transform change now.

**Comment by spencergilbert** [comment]
> Did we lose the change on the source side to `send` instead of `feed` in the pump task? I only see the transform change now.

Indeed, it didn't actually need the change it (changing it broke the topology::backpressure tests), it already had a similar setup of `feed` on a loop and an explicit `flush`:

https://github.com/vectordotdev/vector/blob/dc1ad916ca351792646100fed99201137f68d2e7/src/topology/builder.rs#L161-L164

**Comment by lukesteensen** [comment]
That one I thought was tricky because the flush call is there, but we won't actually hit it until the input stream ends. I think that was related to why `vector tap` wasn't able to read from sources. Are the backpressure test failures just an off-by-one kind of deal? If so, it might be more correct to update the tests.

**Comment by spencergilbert** [comment]
Closing in favor of https://github.com/vectordotdev/vector/pull/10948
